mllib/src/main/scala/org/apache/spark/ml/tree/impl/GradientBoostedTrees.scala

```diff
@@ -329,8 +330,16 @@ private[spark] object GradientBoostedTrees extends Logging {
       //       However, the behavior should be reasonable, though not optimal.
       baseLearnerWeights(m) = learningRate
 
+      if (pre_predError.getStorageLevel != StorageLevel.NONE ){
```

Comment:
Why  would "unpersisting here may simply make it not cache when it needs to be cached" happen? 

First,my experiment indicates  the code above can save the training time from 1.5h to 45min.

The predError RDD  will be cached when a job or action submitted , the only action in the while loop is in `val model = dt.train(data, treeStrategy)`   before  unpersist. 

In first round, predError will not be persist because I didn't persist it outer the loop, and line 333 will do nothing. predError will be transformed  to a new one  and tagged persist  at  line 338 to line 341. 

In the second round, predError RDD has been updated  to a new one  and tagged persist, it will be actually transformed and persisted at line 333  `val model = dt.train(data, treeStrategy)`  . The pre_predError  still refers to the old  predError  so `if (pre_predError.getStorageLevel != StorageLevel.NONE) {`  will do nothing .  Then  pre_predError will refer to predError at `pre_predError=predError` .   Then predError be updated(transform,not immediately) and tagged persist again.

In 3rd round,  the `val model = dt.train(data, treeStrategy)`  submit the job and  use the cached RDD  pre_predError before it unpersist.

Can you explain why you got  "unpersisting here may simply make it not cache when it needs to be cached."?